File: 백준/Gold/1520. 내리막 길/내리막 길.py
import sys
from collections import deque

input = sys.stdin.readline

m, n = map(int, input().split())
route = [list(map(int, input().split())) for i in range(m)]

dp = [[-1]*n for i in range(m)]
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

def count_route(x,y):
    stack = deque()
    stack.append((x,y))
    dp[x][y] = 0

    while stack:
        cx, cy = stack[-1]

        if cx == m-1 and cy == n-1:
            dp[cx][cy] = 1
            stack.pop()
            continue

        allChildProcessed = True
        for i in range(4):
            nx, ny = cx + dx[i], cy + dy[i]
            if 0 <= nx < m and 0 <= ny < n and route[nx][ny] < route[cx][cy]:
                if dp[nx][ny] == -1:
                    stack.append((nx, ny))
                    allChildProcessed = False
                    
        if allChildProcessed:
            dp[cx][cy] = 0
            for i in range(4):
                nx, ny = cx + dx[i], cy + dy[i]
                if 0 <= nx < m and 0 <= ny < n and route[nx][ny] < route[cx][cy]:
                    dp[cx][cy] += dp[nx][ny]
            stack.pop()
    return dp[x][y]

print(count_route(0,0))

File: 백준/Gold/1520. 내리막 길/내리막 길.java
import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

class Main {
    static int m, n;
    static int[][] route, dp;
    static int[] dx = {0, 0, 1, -1};
    static int[] dy = {1, -1, 0, 0};
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(reader.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        route = new int[m][n];
        dp = new int[m][n];
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(reader.readLine());
            for (int j = 0; j < n; j++) {
                route[i][j] = Integer.parseInt(st.nextToken());
                dp[i][j] = -1;
            }
        }
        writer.write(String.valueOf(countRoute(0,0)));
        writer.flush();
        writer.close();
    }

    private static int countRoute(int x, int y) {
        if( x == m-1 && y == n-1 ) {
            return 1;
        }

        if (dp[x][y] != -1) {
            return dp[x][y];
        }

        dp[x][y] = 0;
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if( nx < 0 || nx >= m || ny < 0 || ny >= n ) {
                continue;
            }
            if( route[nx][ny] < route[x][y] ) {
                dp[x][y] += countRoute(nx, ny);
            }
        }
        return dp[x][y];
    }
}